interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -0,0 +1,968 @@
+    }
+  }
+
+  // Iterate over files under this path. We want to get each ".so" files
+  for (const LibraryPath* P : m_Libraries->GetLibraries()) {
+    const std::string LibName = P->GetFullName();
+
+    if (ContainsSymbol(P, mangledName, /*ignore*/
+                       llvm::object::SymbolRef::SF_Undefined)) {
+      m_QueriedLibraries.push_back(*P);
+      return LibName;
+    }
+  }
+
+  if (!searchSystem)
+    return "";
+
+  if (DEBUG > 7)
+    cling::errs() << "Dyld::ResolveSymbol: SearchSystem!\n";
+
+  // Lookup in non-system libraries failed. Expand the search to the system.
+  if (m_FirstRunSysLib) {
+    ScanForLibraries(/* SearchSystemLibraries= */ true);
+    m_FirstRunSysLib = false;
+  }
+
+  for (const LibraryPath* P : m_SysLibraries->GetLibraries()) {
+    const std::string LibName = P->GetFullName();
+    if (ContainsSymbol(P, mangledName, /*ignore*/
+                       llvm::object::SymbolRef::SF_Undefined |
+                       llvm::object::SymbolRef::SF_Weak)) {
+      m_QueriedLibraries.push_back(*P);
+      return LibName;
+    }
+  }
+
+  if (DEBUG > 7)
+    cling::errs() << "Dyld::ResolveSymbol: Search found no match!\n";
+
+  /*
+    if (DEBUG > 7) {
+    cling::errs() << "Dyld::ResolveSymbol: Structs after ResolveSymbol:\n");
+
+    cling::errs() << "Dyld::ResolveSymbol - sPaths:\n");
+    size_t x = 0;
+    for (const auto &item : sPaths.GetPaths())
+    cling::errs() << "Dyld::ResolveSymbol << [" x++ << "]: " << item << "\n";
+
+    cling::errs() << "Dyld::ResolveSymbol - sLibs:\n");
+    x = 0;
+    for (const auto &item : sLibraries.GetLibraries())
+    cling::errs() << "Dyld::ResolveSymbol ["
+    << x++ << "]: " << item->Path << ", "
+    << item->LibName << "\n";
+
+    cling::errs() << "Dyld::ResolveSymbol - sSysLibs:");
+    x = 0;
+    for (const auto &item : sSysLibraries.GetLibraries())
+    cling::errs() << "Dyld::ResolveSymbol ["
+    << x++ << "]: " << item->Path << ", "
+    << item->LibName << "\n";
+
+    Info("Dyld::ResolveSymbol", "- sQueriedLibs:");
+    x = 0;
+    for (const auto &item : sQueriedLibraries)
+    cling::errs() << "Dyld::ResolveSymbol ["
+    << x++ << "]: " << item->Path << ", "
+    << item->LibName << "\n";
+    }
+  */
+
+  return ""; // Search found no match.
+}
+
+// This function isn't referenced outside its translation unit, but it
+// can't use the "static" keyword because its address is used for
+// GetMainExecutable (since some platforms don't support taking the
+// address of main, and some platforms can't implement GetMainExecutable
+// without being given the address of a function in the main executable).
+std::string GetExecutablePath() {
+   // This just needs to be some symbol in the binary; C++ doesn't
+   // allow taking the address of ::main however.
+   return cling::DynamicLibraryManager::getSymbolLocation(&GetExecutablePath);
+}
+
+namespace cling {
+  void
+  DynamicLibraryManager::initializeDyld(llvm::StringSet<> libBlacklist) const {
```

Comment:
Why can this not be part of the constructor? That's safer than some call sequence that must be obeyed. IIUC that would also get rid of `s_IsDyldInitialized`.

Reply:
Alternatively, we could make that part of the ctor, indeed. However, then we require the user to have decided whether to use the dyld or not. This makes it lazy whenever it is needed.